File: cs50-practice/week6/readability.py
# Main Function
def main():
    # Get a text from user
    while True:
        text = input("Text: ")
        if len(text) > 0:
            break

    # Count Letters
    letters = count_letters(text)

    # Count Words
    words = count_words(text)

    # Count Sentences
    sentences = count_sentences(text)

    # Compute Index
    index = compute_index(letters, words, sentences)

    # Print Stats
    print(f"Letters: {letters}")
    print(f"Words: {words}")
    print(f"Sentences: {sentences}")
    print("Index: ", end="")

    # Print Grade Level
    if index < 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {index}")


# Count Letters Function
def count_letters(string):
    letters = 0

    for i in range(len(string)):
        if string[i].isalnum():
            letters += 1

    return letters


# Count Words Function
def count_words(string):
    words = 1

    for i in range(len(string)):
        if string[i] == ' ' and string[i + 1].isalnum():
            words += 1

    return words


# Count Sentences Function
def count_sentences(string):
    sentences = 0

    for i in range(len(string)):
        if string[i] in ['?', '.', '!'] and string[i - 1].isalnum():
            sentences += 1

    return sentences


# Compute Index Function
def compute_index(letters, words, sentences):
    L = letters / words * 100
    S = sentences / words * 100
    index = round(0.0588 * L - 0.296 * S - 15.8)

    return index


if __name__ == "__main__":
    main()

File: cs50-practice/week2/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
int readability_index(int letters, int words, int sentences);

int main(void)
{
    // Get a text from the user
    string text = get_string("Text: ");
    // Count text length
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);
    // Print each
    printf("%i letters\n", letters);
    printf("%i words\n", words);
    printf("%i sentences\n", sentences);
    // Compute the Coleman-Liau index
    int index = readability_index(letters, words, sentences);
    // Print the grade level
    if (index < 1)
    {
        printf("Below Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

int count_letters(string text)
{
    int letter = 0;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (isalnum(text[i]))
        {
            letter++;
        }

    }
    return letter;
}
int count_words(string text)
{
    int word = 0;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (text[i] == ' ')
        {
            word++;
        }
    }
    return word + 1;
}
int count_sentences(string text)
{
    int sentence = 0;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (text[i] == '.' || text[i] == '?' || text[i] == '!')
        {
            sentence++;
        }
    }
    return sentence;
}
int readability_index(int letters, int words, int sentences)
{
    float L = letters / words * 100.0;
    float S = sentences / words * 100.0;
    int index = (int) round(0.0588 * L - 0.296 * S - 15.8);
    return index;
}
